llvm/lib/Analysis/ValueTracking.cpp

```diff
@@ -3043,6 +3043,10 @@ static bool isKnownNonZeroFromOperator(const Operator *I,
     // (X ^ (X != 0)) is non zero
     if (matchOpWithOpEqZero(I->getOperand(0), I->getOperand(1)))
       return true;
+    // X ^ Y != 0 if X != Y.
+    if (isKnownNonEqual(I->getOperand(0), I->getOperand(1), DemandedElts, Q,
+                        Depth))
+      return true;
```

Comment:
Can we make this `return isKnownNonEqual()`? I don't think the fallback path using KnownBits should be able to infer any additional cases, as isKnownNonEqual() also has the KnownBits logic.